docs/design/proxy-support.md

```diff
@@ -0,0 +1,222 @@
+# Supporting package managers proxies in Hermeto
+
+## Background
+
+It might be desirable to pull dependencies through a proxy for a variety of
+reasons (builds reproducibility, additional verification, fetch speedup,
+etc). This means that Hermeto has to support proxying packages for
+individual package managers.
+
+## Proxy usage overview
+
+Most supported package managers (with a notable exception of Yarn v1) make
+use of proxy URL and support standard mechanisms for authentication.  Proxy
+usage is transparent to users as long as they provide correct proxy
+address and authentication credentials. Given that a proxy is correctly
+configured a user would just receive their dependencies as usual with an
+underlying package manager handling everything seamlessly.  Thus a way to
+consume proxy URL and credentials from the environment must be added to
+Hermeto.
+
+A proxy might require authentication or not.  A proxy might exist for just
+some package managers and do not exist for others. This means that each
+individual package manager will need to handle checking the proxy setting
+and authentication setting. Once these settings are known they must be
+made available to the underlying tool used for fetching the actual
+dependencies either via environment or via a configuration option.
+
+
+## Additional considerations
+
+### SBOM enhancement
+
+SBOMs must be marked to indicate whether a proxy was used or not. Currently the
+origin of an artifact is not recorded in a SBOM at all, it is assumed that
+download location is the same as one reported in PURL. With addition of
+proxies support this won't be true anymore. Both supported SBOM formats provide
+means for specifying a component's location: CyclondeDX provides
+[externalReferences][] field for that, and SPDX provides
+[PackageDownloadLocation][]. The most straightforward solution would be to
+include these fields into SBOMs. The downside of this solution is that some
+package managers support specifying multiple proxies in the environment and not
+reporting which specific proxy was used which means that all proxies must be reported.
+This is not the best possible outcome, but in practice would result in
+verification of several proxies instead of one.  Unfortunately
+[PackageDownloadLocation][] does not seem to support multiple locations which
```

Comment:
The fact that it natively doesn't support it is not a show-stopper IMO because at the end of the day it's just a string, same as `GOPROXY` . We can "just happen" to record multiple urls into the field and choose a delimiter ourselves.